phoenix-core/src/main/java/org/apache/phoenix/schema/MetaDataClient.java

```diff
@@ -945,20 +960,40 @@ public MutationState createIndex(CreateIndexStatement statement, byte[][] splits
                     PDataType dataType = MetaDataUtil.getViewIndexIdDataType();
                     ColumnName colName = ColumnName.caseSensitiveColumnName(MetaDataUtil.getViewIndexIdColumnName());
                     allPkColumns.add(new Pair<ColumnName, SortOrder>(colName, SortOrder.getDefault()));
-                    columnDefs.add(FACTORY.columnDef(colName, dataType.getSqlTypeName(), false, null, null, false, SortOrder.getDefault()));
+                    columnDefs.add(FACTORY.columnDef(colName, dataType.getSqlTypeName(), false, null, null, false, SortOrder.getDefault(), null));
                 }
                 // First columns are the indexed ones
-                for (Pair<ColumnName, SortOrder> pair : indexedPkColumns) {
-                    ColumnName colName = pair.getFirst();
-                    PColumn col = resolver.resolveColumn(null, colName.getFamilyName(), colName.getColumnName()).getColumn();
-                    unusedPkColumns.remove(col);
-                    // Ignore view constants for updatable views as we don't need these in the index
-                    if (col.getViewConstant() == null) {
-                        PDataType dataType = IndexUtil.getIndexColumnDataType(col);
-                        colName = ColumnName.caseSensitiveColumnName(IndexUtil.getIndexColumnName(col));
-                        allPkColumns.add(new Pair<ColumnName, SortOrder>(colName, pair.getSecond()));
-                        columnDefs.add(FACTORY.columnDef(colName, dataType.getSqlTypeName(), col.isNullable(), col.getMaxLength(), col.getScale(), false, SortOrder.getDefault()));
-                    }
+                for (Pair<ParseNode, SortOrder> pair : indexParseNodeAndSortOrderList) {
+                	ParseNode parseNode = pair.getFirst();
+                	ColumnName colName = null;
+                	// if it is a column 
+                	if (parseNode instanceof ColumnParseNode) {
+                		ColumnParseNode colParseNode = (ColumnParseNode)parseNode;
+	                    PColumn col = resolver.resolveColumn(null, colParseNode.getTableName(), colParseNode.getName()).getColumn();
+	                    unusedPkColumns.remove(col);
+	                    // Ignore view constants for updatable views as we don't need these in the index
+	                    if (col.getViewConstant() != null) 
+	                    	continue;
+	                    colName = ColumnName.caseSensitiveColumnName(IndexUtil.getIndexColumnName(col));
+                	}
+            	    // compile the parseNode to get an expression
```

Comment:
You should normalize the parseNode too. This puts constants on the RHS as expected among other things. You can do this like this:

    parseNode = StatementNormalizer.normalize(parseNode, resolver);

A good test is to use a constant on the RHS and it'll be moved to the LHS (like in a CASE WHEN statement).